matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/identity/DefaultIdentityService.kt

```diff
@@ -0,0 +1,345 @@
+import javax.inject.Inject
+import javax.net.ssl.HttpsURLConnection
+
+@SessionScope
+internal class DefaultIdentityService @Inject constructor(
+        private val identityStore: IdentityStore,
+        private val getOpenIdTokenTask: GetOpenIdTokenTask,
+        private val identityBulkLookupTask: IdentityBulkLookupTask,
+        private val identityRegisterTask: IdentityRegisterTask,
+        private val identityPingTask: IdentityPingTask,
+        private val identityDisconnectTask: IdentityDisconnectTask,
+        private val identityRequestTokenForBindingTask: IdentityRequestTokenForBindingTask,
+        @Unauthenticated
+        private val unauthenticatedOkHttpClient: Lazy<OkHttpClient>,
+        @AuthenticatedIdentity
+        private val okHttpClient: Lazy<OkHttpClient>,
+        private val retrofitFactory: RetrofitFactory,
+        private val coroutineDispatchers: MatrixCoroutineDispatchers,
+        private val updateUserAccountDataTask: UpdateUserAccountDataTask,
+        private val bindThreePidsTask: BindThreePidsTask,
+        private val submitTokenForBindingTask: IdentitySubmitTokenForBindingTask,
+        private val unbindThreePidsTask: UnbindThreePidsTask,
+        private val identityApiProvider: IdentityApiProvider,
+        private val accountDataDataSource: AccountDataDataSource,
+        private val homeServerCapabilitiesService: HomeServerCapabilitiesService,
+        private val sessionParams: SessionParams
+) : IdentityService {
+
+    private val lifecycleOwner: LifecycleOwner = LifecycleOwner { lifecycleRegistry }
+    private val lifecycleRegistry: LifecycleRegistry = LifecycleRegistry(lifecycleOwner)
+
+    private val listeners = mutableSetOf<IdentityServiceListener>()
+
+    fun start() {
+        lifecycleRegistry.currentState = Lifecycle.State.STARTED
+        // Observe the account data change
+        accountDataDataSource
+                .getLiveAccountDataEvent(UserAccountData.TYPE_IDENTITY_SERVER)
+                .observeNotNull(lifecycleOwner) {
+                    notifyIdentityServerUrlChange(it.getOrNull()?.content?.toModel<IdentityServerContent>()?.baseUrl)
+                }
+
+        // Init identityApi
+        updateIdentityAPI(identityStore.getIdentityData()?.identityServerUrl)
+    }
+
+    private fun notifyIdentityServerUrlChange(baseUrl: String?) {
+        // This is maybe not a real change (echo of account data we are just setting)
+        if (identityStore.getIdentityData()?.identityServerUrl == baseUrl) {
+            Timber.d("Echo of local identity server url change, or no change")
+        } else {
+            // Url has changed, we have to reset our store, update internal configuration and notify listeners
+            identityStore.setUrl(baseUrl)
+            updateIdentityAPI(baseUrl)
+            listeners.toList().forEach { tryThis { it.onIdentityServerChange() } }
+        }
+    }
+
+    fun stop() {
+        lifecycleRegistry.currentState = Lifecycle.State.DESTROYED
```

Comment:
So this will destroy lifecycle owner and release life data strong reference to observers?